src/libraries/System.Text.Json/Common/JsonSeparatorNamingPolicy.cs

```diff
@@ -21,149 +28,154 @@ public sealed override string ConvertName(string name)
+                            case SeparatorState.SpaceSeparator:
+                                // An uppercase letter following a sequence of lowercase letters or spaces
+                                // denotes the start of a new grouping: emit a separator character.
+                                WriteChar(separator, ref destination);
+                                break;
+
+                            case SeparatorState.UppercaseLetter:
+                                // We are reading through a sequence of two or more uppercase letters.
+                                // Uppercase letters are grouped together with the exception of the
+                                // final letter, assuming it is followed by lowercase letters.
+                                // For example, the value 'XMLReader' should render as 'xml_reader',
+                                // however 'SHA512Hash' should render as 'sha512-hash'.
+                                if (i + 1 < chars.Length && char.IsLower(chars[i + 1]))
+                                {
+                                    WriteChar(separator, ref destination);
+                                }
+                                break;
+
+                            default:
+                                Debug.Fail($"Unexpected state {state}");
+                                break;
                         }
-                    }
 
-                    ExpandBuffer(ref result);
-                }
+                        if (lowercase)
+                        {
+                            current = char.ToLowerInvariant(current);
+                        }
 
-                if (resultUsedLength != 0)
-                {
-                    result[resultUsedLength] = _separator;
-                    resultUsedLength += 1;
-                }
+                        WriteChar(current, ref destination);
+                        state = SeparatorState.UppercaseLetter;
+                        break;
 
-                resultUsedLength += written;
-            }
+                    case UnicodeCategory.LowercaseLetter:
+                    case UnicodeCategory.DecimalDigitNumber:
 
-            int first = 0;
-            ReadOnlySpan<char> chars = name.AsSpan();
-            CharCategory previousCategory = CharCategory.Boundary;
+                        if (state is SeparatorState.SpaceSeparator)
+                        {
+                            // Normalize preceding spaces to one separator.
+                            WriteChar(separator, ref destination);
+                        }
 
-            for (int index = 0; index < chars.Length; index++)
-            {
-                char current = chars[index];
-                UnicodeCategory currentCategoryUnicode = char.GetUnicodeCategory(current);
+                        if (!lowercase)
+                        {
+                            current = char.ToUpperInvariant(current);
```

Comment:
>ToUpperInvariant(current); [](http://example.com/codeflow?start=43&length=26)

do we need to upper case when having DecimalDigitNumber?

Reply:
Only letters can have casing.